# Assignment: Data Wrangling

### Reading material: `tidy_data.pdf`

**Q1.** This question provides some practice cleaning variables which have common problems.
1. For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing?

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks (Hint: `GSAF5.xls`).

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?

---

## *Phase 1: Solution without AI

Complete all questions in the following cells without generative AI. Then commit and push the
notebook before beginning Phase 2.

**Declaration:** I completed this version without generative AI.

**Student(s):** Raffaella Alpaca

**Date:** 09/02/2026

Question 1 - Part 1

In [42]:
# Your Phase 1 solution to the problem goes here.
import pandas as pd
import matplotlib.pyplot as plt
airbnb = pd.read_csv('https://raw.githubusercontent.com/sds3001/A01-data-wrangling/main/data/airbnb_hw.csv')

#shows first 5 rows to see the columns
airbnb.head()
#to show price column
airbnb['Price'].head()

,Price
0,145
1,37
2,28
3,199
4,549


In [17]:
#to inspect the bottom part
airbnb['Price'].tail()

,Price
30473,300
30474,125
30475,80
30476,35
30477,80


In [18]:
airbnb['Price'].dtype

dtype('O')

In [23]:
#looking for prices that have commas
airbnb[airbnb['Price'].str.contains(',')]

,Host Id,Host Since,Name,Neighbourhood,Property Type,Review Scores Rating (bin),Room Type,Zipcode,Beds,Number of Records,Number Of Reviews,Price,Review Scores Rating
101,8730,2/27/2009,Comfort & Convenience,Manhattan,Condominium,100.0,Entire home/apt,10003.0,6.0,1,4,"1,990",100.0
263,34915,8/30/2009,Luxurious 2-Floor Manhattan Mansion,Manhattan,Apartment,NaN,Entire home/apt,10075.0,2.0,1,0,"1,000",NaN
764,112879,4/23/2010,Modern Luxury Meets Old Money Charm,Brooklyn,Apartment,NaN,Entire home/apt,11206.0,5.0,1,0,"1,200",NaN
1272,214148,8/26/2010,"LARGE, COMFY 1BDR W/CHARACTER!!!",Brooklyn,Apartment,90.0,Entire home/apt,11216.0,1.0,1,30,"1,000",92.0
1275,213266,8/26/2010,Beautiful 1 Bedroom in Nolita/Soho,Manhattan,Apartment,75.0,Entire home/apt,10012.0,2.0,1,2,"5,000",76.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
28951,37155810,6/30/2015,Central Park Luxury : 115916,Manhattan,Apartment,NaN,Entire home/apt,10019.0,2.0,1,0,"3,390",NaN
28952,37155810,6/30/2015,Central Park South Place : 112756,Manhattan,Apartment,NaN,Entire home/apt,10019.0,1.0,1,0,"1,356",NaN
28953,37155810,6/30/2015,Central Park South Views : 112754,Manhattan,Apartment,NaN,Entire home/apt,10019.0,1.0,1,0,"2,599",NaN
28985,37241813,7/1/2015,VERY Spacious 3BR in East Village,Manhattan,Apartment,NaN,Entire home/apt,10009.0,3.0,1,0,"2,000",NaN


In [29]:
#price values that are problematic
prices_w_commas = airbnb[airbnb['Price'].str.contains(',')]
prices_w_commas['Price'].head()

,Price
101,"1,990"
263,"1,000"
764,"1,200"
1272,"1,000"
1275,"5,000"


In [34]:
#removimg commas
airbnb['Price'] = airbnb['Price'].str.replace(',', '')
#checking again
airbnb['Price'].head()

,Price
0,145
1,37
2,28
3,199
4,549


In [35]:
airbnb['Price'].dtype

dtype('O')

In [40]:
#dtype is still an object
airbnb['Price'] = pd.to_numeric(airbnb['Price'])
airbnb['Price'].dtype

dtype('int64')

In [41]:
airbnb['Price'].head()

,Price
0,145
1,37
2,28
3,199
4,549


Prices over 999 contained commas such as 1,000 or 1,990. This caused the price column to be stores as object instead of being a numeric type. I removed the commas and converted the column to numeric values

Question 2 - Part 2

In [45]:
#loading dataset

police = pd.read_csv('https://raw.githubusercontent.com/sds3001/A01-data-wrangling/main/data/mn_police_use_of_force.csv')

In [46]:
police.head()

,response_datetime,problem,is_911_call,primary_offense,subject_injury,force_type,force_type_action,race,sex,age,type_resistance,precinct,neighborhood
0,2016/01/01 00:47:36,Assault in Progress,Yes,DASLT1,NaN,Bodily Force,Body Weight to Pin,Black,Male,20.0,Tensed,1,Downtown East
1,2016/01/01 02:19:34,Fight,No,DISCON,NaN,Chemical Irritant,Personal Mace,Black,Female,27.0,Verbal Non-Compliance,1,Downtown West
2,2016/01/01 02:19:34,Fight,No,DISCON,NaN,Chemical Irritant,Personal Mace,White,Female,23.0,Verbal Non-Compliance,1,Downtown West
3,2016/01/01 02:28:48,Fight,No,PRIORI,NaN,Chemical Irritant,Crowd Control Mace,Black,Male,20.0,Commission of Crime,1,Downtown West
4,2016/01/01 02:28:48,Fight,No,PRIORI,NaN,Chemical Irritant,Crowd Control Mace,Black,Male,20.0,Commission of Crime,1,Downtown West


In [47]:
police['subject_injury'].value_counts(dropna=False)

,count
subject_injury,
NaN,9848
Yes,1631
No,1446


In [48]:
#before deciding what to do with missing values, im gonna check if related to force_type
pd.crosstab(police['subject_injury'], police['force_type'])

force_type,Baton,Bodily Force,Chemical Irritant,Firearm,Gun Point Display,Improvised Weapon,Less Lethal Projectile,Police K9 Bite,Taser
subject_injury,,,,,,,,,
No,0,1093,131,2,33,34,1,2,150
Yes,2,1286,41,0,44,40,2,44,172


In [52]:
#making a temp version for missing values
injury_check = police['subject_injury'].fillna('Missing')

In [54]:
pd.crosstab(injury_check, police['force_type'])

force_type,Baton,Bodily Force,Chemical Irritant,Firearm,Gun Point Display,Improvised Weapon,Less Lethal,Less Lethal Projectile,Maximal Restraint Technique,Police K9 Bite,Taser
subject_injury,,,,,,,,,,,
Missing,2,7051,1421,0,27,74,87,0,170,31,985
No,0,1093,131,2,33,34,0,1,0,2,150
Yes,2,1286,41,0,44,40,0,2,0,44,172


In [56]:
#calc the overall proportion of subject_injury that is missing
police['subject_injury'].isna().mean()

np.float64(0.7619342359767892)

Around 76% of the subject_injury values are missing. This is concerning because most of the data is not available , so its hard to make a promising conclusion about injuries.Also the missing values appear a lot for some force tyoes, so the missing data may not be fully random.

In [60]:
#cleaning column
police['injury_clean']= police['subject_injury']

In [61]:
police['injury_clean'].value_counts(dropna=False)

,count
injury_clean,
NaN,9848
Yes,1631
No,1446


The recorded subject_injury values were already labeled as yes and no. I decided to keeo the missing values as missing because it doesnt mean that no injuries occurred.

In [62]:
#cross tab using the clean column
pd.crosstab(police['injury_clean'], police['force_type'])

force_type,Baton,Bodily Force,Chemical Irritant,Firearm,Gun Point Display,Improvised Weapon,Less Lethal Projectile,Police K9 Bite,Taser
injury_clean,,,,,,,,,
No,0,1093,131,2,33,34,1,2,150
Yes,2,1286,41,0,44,40,2,44,172


The cross tab shows that injuries are common for some tyoe force than others. Bodily force has many injury cases, while on the other hand, other force types have fewer.

## *Phase 2: AI-assisted Revision
Keep Phase 1 frozen.

- **Phase 1 commit SHA ID:** [To be provided]
- **AI tool and model used:** [To be provided]

### *Prompts used

1. [Paste prompt]
2. [Paste follow-up prompt]

### *AI-proposed changes


[Paste the AI's concise numbered list verbatim.]

1. ...
2. ...

### *Evaluation of AI suggestions


| Change | Decision | Reason and verification |
|---|---|---|
| 1 | Accept / modify / reject | ... |
| 2 | Accept / modify / reject | ... |

### *AI-assisted solution in full
After the evaluation of AI suggestions, incorporate the accepted revisions into your Phase 1 solution and provide the final Phase 2 solution in full in the following cells.

In [ ]:
# Your Phase 2 solution to the problem goes here.


### *Reflection on AI assistance
In your own words without generative AI.

[In 150–300 words, describe the main improvements, AI mistakes or limitations, how the final solution was verified, and any remaining concerns.]